### PR TITLE
Pivotal ID # 168727325: Project Submission Relationship

### PR DESCRIPTION
**Pull request by jhoanmanuelms**
- Validate that the project has the corresponding access tag
- Add access tag for the test project
- Add an endpoint and a web client method to create projects
- The corresponding access tag for the project will be automatically created
- Use the new methods in the projects API test

TODO In Next PR
- Some classes require unit testing
- Validate only superusers are able to submit projects
- The integration tests that use projects should validate tags

### CODE REVIEW

File: submission/persistence/src/main/kotlin/ac/uk/ebi/biostd/persistence/integration/PersistenceContextImpl.kt [hunk 1]
```diff
@@ -60,6 +62,12 @@ open class PersistenceContextImpl(
         }
     }
 
+    override fun saveAccessTag(accessTag: String) {
+        accessTagsDataRepository.save(AccessTag(name = accessTag))
```
Comment (Juan-EBI) [inline submission/persistence/src/main/kotlin/ac/uk/ebi/biostd/persistence/integration/PersistenceContextImpl.kt]: inliner

File: submission/persistence/src/main/kotlin/ac/uk/ebi/biostd/persistence/service/SubmissionRepository.kt [hunk 2]
```diff
@@ -34,6 +34,7 @@ class SubmissionRepository(
         submissionRepository.findByTypeAndAccNo(Project.value, tags.map { it.name })
             .map { submissionDbMapper.toSubmission(it) }
 
+    // TODO this method should return the submission model, not the persistence object
```
Comment (Juan-EBI) [inline submission/persistence/src/main/kotlin/ac/uk/ebi/biostd/persistence/service/SubmissionRepository.kt]: do not get this TODO, this one is a spring data repository

File: submission/submitter/src/main/kotlin/ac/uk/ebi/biostd/submission/processors/ProjectProcessor.kt [hunk 3]
```diff
@@ -0,0 +1,11 @@
+package ac.uk.ebi.biostd.submission.processors
+
+import ebi.ac.uk.model.ExtendedSubmission
+import ebi.ac.uk.model.extensions.addAccessTag
+import ebi.ac.uk.persistence.PersistenceContext
+
+class ProjectProcessor : SubmissionProcessor {
```
Comment (Juan-EBI) [inline submission/submitter/src/main/kotlin/ac/uk/ebi/biostd/submission/processors/ProjectProcessor.kt]: Is this only applied to the project? as it implements SubmissionProcessor I would expect is applied to every submission and it does look correct them

File: submission/submitter/src/main/kotlin/ac/uk/ebi/biostd/submission/validators/ProjectValidator.kt [hunk 4]
```diff
@@ -1,12 +1,19 @@
...
+import ac.uk.ebi.biostd.submission.exceptions.ProjectAlreadyExistingException
+import ebi.ac.uk.base.ifTrue
 import ebi.ac.uk.model.ExtendedSubmission
-import ebi.ac.uk.model.extensions.attachTo
 import ebi.ac.uk.persistence.PersistenceContext
 
 class ProjectValidator : SubmissionValidator {
     override fun validate(submission: ExtendedSubmission, context: PersistenceContext) {
-        submission.attachTo?.let { context.getSubmission(it) ?: throw InvalidProjectException(it) }
+        context.isNew(submission.accNo).not().ifTrue {
```
Comment (Juan-EBI) [inline submission/submitter/src/main/kotlin/ac/uk/ebi/biostd/submission/validators/ProjectValidator.kt]: we may need better `ifFalse` extension

File: submission/submitter/src/main/kotlin/ac/uk/ebi/biostd/submission/validators/SubmissionProjectValidator.kt [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+        submission.attachTo?.let {
+            validateProject(it, context)
+            validateProjectAccessTag(it, submission, context)
+        }
+    }
+
+    private fun validateProject(project: String, context: PersistenceContext) =
+        context.getSubmission(project) ?: throw InvalidProjectException(project)
+
+    private fun validateProjectAccessTag(project: String, submission: ExtendedSubmission, context: PersistenceContext) =
```
Comment (Juan-EBI) [inline submission/submitter/src/main/kotlin/ac/uk/ebi/biostd/submission/validators/SubmissionProjectValidator.kt]: what is the purpose of this validation? the idea AFAIK is that the system will create access tag for the project. Users should not use access tags.

File: submission/persistence/src/main/kotlin/ac/uk/ebi/biostd/persistence/integration/PersistenceContextImpl.kt (hunk 1 above)
Comment (jhoanmanuelms) [reply submission/persistence/src/main/kotlin/ac/uk/ebi/biostd/persistence/integration/PersistenceContextImpl.kt]: This is a unit return type

File: submission/persistence/src/main/kotlin/ac/uk/ebi/biostd/persistence/service/SubmissionRepository.kt (hunk 2 above)
Comment (jhoanmanuelms) [reply submission/persistence/src/main/kotlin/ac/uk/ebi/biostd/persistence/service/SubmissionRepository.kt]: In all the other methods, we return the submission model object but in this case, we return the persistence object. This method should also use the Submission Mapper and return a Submission model object

File: submission/submitter/src/main/kotlin/ac/uk/ebi/biostd/submission/processors/ProjectProcessor.kt (hunk 3 above)
Comment (jhoanmanuelms) [reply submission/submitter/src/main/kotlin/ac/uk/ebi/biostd/submission/processors/ProjectProcessor.kt]: Well, the project is a submission. In this case, this is a processor for the projects which follows the same validation/processing approach.

Maybe we could rename the interface as `SubmitterProcessor` and `SubmitterValidator`?

File: submission/submitter/src/main/kotlin/ac/uk/ebi/biostd/submission/validators/ProjectValidator.kt (hunk 4 above)
Comment (jhoanmanuelms) [reply submission/submitter/src/main/kotlin/ac/uk/ebi/biostd/submission/validators/ProjectValidator.kt]: Done

File: submission/submitter/src/main/kotlin/ac/uk/ebi/biostd/submission/validators/SubmissionProjectValidator.kt (hunk 5 above)
Comment (jhoanmanuelms) [reply submission/submitter/src/main/kotlin/ac/uk/ebi/biostd/submission/validators/SubmissionProjectValidator.kt]: This validation is intended for projects that were manually created and the access tag might not exist.

File: submission/persistence/src/main/kotlin/ac/uk/ebi/biostd/persistence/service/SubmissionRepository.kt (hunk 2 above)
Comment (Juan-EBI) [reply submission/persistence/src/main/kotlin/ac/uk/ebi/biostd/persistence/service/SubmissionRepository.kt]: ohh got it I was thinking this was `SubmissionDataRepository` and not `SubmissionRepository`

**Pull request merged**